src/pke/lib/scheme/ckksrns/ckksrns-pke.cpp

```diff
@@ -0,0 +1,95 @@
+//==================================================================================
+// BSD 2-Clause License
+//
+// Copyright (c) 2014-2022, NJIT, Duality Technologies Inc. and other contributors
+//
+// All rights reserved.
+//
+// Author TPOC: contact@openfhe.org
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this
+//    list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice,
+//    this list of conditions and the following disclaimer in the documentation
+//    and/or other materials provided with the distribution.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//==================================================================================
+
+/*
+CKKS implementation. If NOISE_FLOODING_DECRYPT is set, we flood the decryption bits with noise.
+ */
+
+#define PROFILE
+
+#include "cryptocontext.h"
+#include "scheme/ckksrns/ckksrns-cryptoparameters.h"
+#include "scheme/ckksrns/ckksrns-pke.h"
+
+namespace lbcrypto {
+
+DecryptResult PKECKKSRNS::Decrypt(ConstCiphertext<DCRTPoly> ciphertext, const PrivateKey<DCRTPoly> privateKey,
+                                  NativePoly* plaintext) const {
+    const auto cryptoParams = std::dynamic_pointer_cast<CryptoParametersCKKSRNS>(ciphertext->GetCryptoParameters());
+    const std::vector<DCRTPoly>& cv = ciphertext->GetElements();
+    DCRTPoly b                      = DecryptCore(cv, privateKey);
+    if (cryptoParams->GetDecryptionNoiseMode() == NOISE_FLOODING_DECRYPT &&
+        cryptoParams->GetExecutionMode() == EXEC_EVALUATION) {
+        auto dgg = cryptoParams->GetFloodingDiscreteGaussianGenerator();
+        DCRTPoly noise(dgg, cv[0].GetParams(), Format::EVALUATION);
+        b += noise;
+    }
+
+    b.SetFormat(Format::COEFFICIENT);
+    const size_t sizeQl = b.GetParams()->GetParams().size();
+
+    if (sizeQl != 1)
```

Comment:
@yspolyakov Should we add logic here to take care of the case where sizeQl > 1? The code for Decrypt with Poly instead of NativePoly has this in place.

Reply:
No, NativePoly assumes there is only one RNS limb left. We never set the modulus size to less than 30 bits in CKKS (the precision would be too small). So I don't think any change is needed.

Reply:
I consulted with Dmitriy, and we left it as one exception, with a more verbose error message, where it prints what sizeQl is.